contracts/libs/data-structures/SparseMerkleTree.sol

```diff
@@ -0,0 +1,578 @@
+                currentDepth_ + 1
+            );
+
+            if (newLeafBitAtDepth_) {
+                // go right
+                newNodeMiddle_ = Node({
+                    nodeType: NodeType.MIDDLE,
+                    childLeft: ZERO_IDX,
+                    childRight: uint64(nextNodeId_),
+                    nodeHash: HASH_ZERO,
+                    index: HASH_ZERO,
+                    value: HASH_ZERO
+                });
+            } else {
+                // go left
+                newNodeMiddle_ = Node({
+                    nodeType: NodeType.MIDDLE,
+                    childLeft: uint64(nextNodeId_),
+                    childRight: ZERO_IDX,
+                    nodeHash: HASH_ZERO,
+                    index: HASH_ZERO,
+                    value: HASH_ZERO
+                });
+            }
+
+            return _setNode(tree, newNodeMiddle_);
+        }
+
+        uint256 newLeafId = _setNode(tree, newLeaf_);
+
+        if (newLeafBitAtDepth_) {
+            newNodeMiddle_ = Node({
+                nodeType: NodeType.MIDDLE,
+                childLeft: uint64(oldLeafId_),
+                childRight: uint64(newLeafId),
+                nodeHash: HASH_ZERO,
+                index: HASH_ZERO,
+                value: HASH_ZERO
+            });
+        } else {
+            newNodeMiddle_ = Node({
+                nodeType: NodeType.MIDDLE,
+                childLeft: uint64(newLeafId),
+                childRight: uint64(oldLeafId_),
+                nodeHash: HASH_ZERO,
+                index: HASH_ZERO,
+                value: HASH_ZERO
+            });
+        }
+
+        return _setNode(tree, newNodeMiddle_);
+    }
+
+    function _setNode(SMT storage tree, Node memory node_) private returns (uint256) {
+        node_.nodeHash = _getNodeHash(tree, node_);
+
+        uint256 newSize_ = ++tree.nodesCount;
+        tree.nodes[newSize_] = node_;
+
+        return newSize_;
+    }
+
+    // TODO: Explain why Empty nodes cannot be passed to the function
+    function _getNodeHash(SMT storage tree, Node memory node_) private view returns (bytes32) {
+        function(bytes32, bytes32) view returns (bytes32) hash2_ = tree.isCustomHasherSet
+            ? tree.hash2
+            : _hash2;
+        function(bytes32, bytes32, bytes32) view returns (bytes32) hash3_ = tree.isCustomHasherSet
+            ? tree.hash3
+            : _hash3;
+
+        if (node_.nodeType == NodeType.LEAF) {
+            return hash3_(node_.index, node_.value, bytes32(uint256(1)));
+        }
+
+        return hash2_(tree.nodes[node_.childLeft].nodeHash, tree.nodes[node_.childRight].nodeHash);
+    }
+
+    function _hash2(bytes32 a, bytes32 b) private pure returns (bytes32 result) {
+        assembly {
+            mstore(0, a)
+            mstore(0x20, b)
+
+            result := keccak256(0, 0x40)
+        }
+    }
+
+    function _hash3(bytes32 a, bytes32 b, bytes32 c) private pure returns (bytes32 result) {
+        assembly {
```

Comment:
Maybe add a dev comment stating "we are not updating the free memory pointer here due to the temporary nature of hash arguments"